pkg/provisioner/ironic/ironic_test.go

```diff
@@ -20,58 +26,110 @@ func init() {
-			Image: &metal3v1alpha1.Image{
-				URL: "not-empty",
-			},
-			Online:          true,
-			HardwareProfile: "libvirt",
-			RootDeviceHints: &metal3v1alpha1.RootDeviceHints{
-				DeviceName:         "userd_devicename",
-				HCTL:               "1:2:3:4",
-				Model:              "userd_model",
-				Vendor:             "userd_vendor",
-				SerialNumber:       "userd_serial",
-				MinSizeGigabytes:   40,
-				WWN:                "userd_wwn",
-				WWNWithExtension:   "userd_with_extension",
-				WWNVendorExtension: "userd_vendor_extension",
-				Rotational:         &rotational,
-			},
+		{
+			name:      "NoDriversLoaded",
+			ironic:    newMockServer(6385),
+			inspector: newMockServer(5050),
+
+			expectedIronicCalls:  "localhost:6385/v1;localhost:6385/v1/drivers;",
+			expectedRequeueAfter: 10 * time.Second,
 		},
-		Status: metal3v1alpha1.BareMetalHostStatus{
-			Provisioning: metal3v1alpha1.ProvisionStatus{
-				ID: "provisioning-id",
-				// Place the hints in the status field to pretend the
-				// controller has already reconciled partially.
-				RootDeviceHints: &metal3v1alpha1.RootDeviceHints{
-					DeviceName:         "userd_devicename",
-					HCTL:               "1:2:3:4",
-					Model:              "userd_model",
-					Vendor:             "userd_vendor",
-					SerialNumber:       "userd_serial",
-					MinSizeGigabytes:   40,
-					WWN:                "userd_wwn",
-					WWNWithExtension:   "userd_with_extension",
-					WWNVendorExtension: "userd_vendor_extension",
-					Rotational:         &rotational,
-				},
-			},
-			HardwareProfile: "libvirt",
+		{
+			name:      "IronicDown",
+			inspector: newMockServer(5050),
+
+			expectedError: "Get http://localhost:6385/v1: dial tcp .*:6385: connect: connection refused",
+		},
+		{
+			name:   "InspectorDown",
+			ironic: newMockServer(6385).addDrivers(),
+
+			expectedIronicCalls: "localhost:6385/v1;localhost:6385/v1/drivers;",
+			expectedError:       "Get http://localhost:5050/v1: dial tcp .*:5050: connect: connection refused",
+		},
+		{
+			name:      "IronicNotOk",
+			ironic:    newMockServer(6385).setInternalError(),
+			inspector: newMockServer(5050),
+
+			expectedIsReady:      false,
+			expectedRequeueAfter: 10 * time.Second,
+
+			expectedIronicCalls: "localhost:6385/v1;",
+		},
+		{
+			name:      "InspectorNotOk",
+			ironic:    newMockServer(6385).addDrivers(),
+			inspector: newMockServer(5050).setInternalError(),
+
+			expectedIsReady:      false,
+			expectedRequeueAfter: 10 * time.Second,
+
+			expectedIronicCalls:    "localhost:6385/v1;localhost:6385/v1/drivers;",
+			expectedInspectorCalls: "localhost:5050/v1;",
 		},
 	}
 
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			if tc.ironic != nil {
+				tc.ironic.start()
+				defer tc.ironic.stop()
+			}
+
+			if tc.inspector != nil {
+				tc.inspector.start()
+				defer tc.inspector.stop()
+			}
+
+			prov, err := newProvisioner(makeHost(), bmc.Credentials{}, nil)
```

Comment:
If we're going to go to the trouble of creating a separate type to check the dependencies, should these tests invoke that type instead of the provisioner? Or can we collapse that type to a few private functions on the provisioner and call them inline from the provisioner's `IsReady()` function?

Reply:
Separate type in this case helps mostly for what regards readability and maintenance, as previously discussed about moving the new logic into a new type for a cleaner approach (effectively ironic type it's already quite big). Leaving the tests on the provisioner type ensure that the provisioner behaviour continues to be the expected one regardless the underlying implementation.